prism/prism.py

```diff
@@ -3,79 +3,375 @@
 
 The Prism API provides a flexible, secure and scalable way to load data into
 Workday Prism Analytics.
+
+DocString style: https://www.sphinx-doc.org/en/master/usage/extensions/example_numpy.html
 """
 
 import logging
 import json
-import random
 import requests
-
-# set up basic logging
-logger = logging.getLogger()
-handler = logging.StreamHandler()
-formatter = logging.Formatter("%(asctime)s %(levelname)s: %(message)s", "%Y-%m-%d %H:%M:%S")
-handler.setFormatter(formatter)
+import time
+import os
+import sys
+import uuid
+import gzip
+import inspect
+import copy
+
+from urllib import parse as urlparse
+
+# Default a logger - the default may be re-configured in the set_logging method.
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.WARNING)
+
+# writing to stdout only...
+handler = logging.StreamHandler(sys.stdout)
+handler.setLevel(logging.WARNING)
+log_format = logging.Formatter("%(asctime)s %(levelname)s: %(message)s", "%Y-%m-%d %H:%M:%S")
+handler.setFormatter(log_format)
 logger.addHandler(handler)
-logger.setLevel(logging.INFO)
 
 
-def load_schema(filename):
-    """Load schema from a JSON file.
+def set_logging(log_file=None, log_level="INFO"):
+    """
+
+    :param log_file:
+    :param log_level:
+    :return:
+    """
+    # Resolve the log level - default to info if empty or invalid.
+    if log_level is None:
+        set_level = logging.INFO
+    else:
+        # Make sure the caller gave us a valid "name" (INFO/DEBUG/etc) for logging level.
+        if hasattr(logging, log_level):
+            set_level = getattr(logging, log_level)
+        else:
+            set_level = logging.INFO
+
+    # If no file was specified, simply loop over any handlers and
+    # set the logging level.
+    if log_file is None:
+        logger.setLevel(set_level)
+
+        for log_handler in logger.handlers:
+            log_handler.setLevel(set_level)
+    else:
+        # Setup logging for CLI operations.
+        for log_handler in logger.handlers:
+            logger.removeHandler(log_handler)
+
+        logger.setLevel(set_level)
+
+        # Create a handler as specified by the user (or defaults)
+        fh = logging.FileHandler(log_file)
+        fh.setLevel(set_level)
+
+        # create formatter and add it to the handlers
+        formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
+        fh.setFormatter(formatter)
+
+        logger.addHandler(fh)
+
+    logger.debug(f"set log level: {set_level}")
+
+
+def log_elapsed(msg, timedelta):
+    """Log the elapsed time of a get/post/put/patch HTTP operation."""
+    elapsed = timedelta.total_seconds()
+    logger.debug(f"{msg}: elapsed {elapsed:.5f}")
+
+
+def buckets_gen_name():
+    bucket_name = "cli_" + uuid.uuid4().hex
```

Comment:
Per advice for the Workday-on-Workday team, they would like the bucket name to start with `prism_python_` so they can audit how this library is being used.  `cli_` feels too generic and not clear that it relates to this interface of the Prism API.

Reply:
Done.